### ISSUE TITLE
Strict mode is wrong

### ISSUE THREAD

**Issue by josephjclark** [OpenFn/kit#233]
I learned this morning that I've fundamentally misunderstood the handling and usage of `state.data`, and as a result, strict mode is wrong.

I thought:

The runtime will manage various bits of state (`config`, `references`), but `data` is for your application data. `state.data` is the thing you want forwarded on to the next job, it's the thing you want in your final result.

I also knew that adaptors like `http` will scribble on `data` (ie, the response). Which I never liked because it erodes the idea that `data` is a safe space for the user.

But all this is incorrect. In fact the opposite is true:

The runtime will manage the `configuration`, `references` and `data` properties of `state`. The rest of the object is left alone for the user to do what they want with.

Adaptors don't modify `state.data` , they _replace_ it (via `composeNextState`)

The idea of strict mode was to establish new, possibly breaking, conventions for the v2 runtime. This includes encouraging users to lean more heavily on `state.data` and take ownership of it, but it turns out that doesn't make any sense.

## Options

* Remove strict mode entirely (this may be a wise short-term measure)
* Keep strict mode but change its meaning. Maybe strict mode returns state apart from `configuration`, `references`, `data` - because that's all  transient stuff needed for the job.

I kinda wish we had a property of state which represents the result, the final or persistent state, the thing that SHOULD get passed to the next job. A part of state which we promise we won't touch or interfere with, since we're a bit loosey goosey about a small number of keys.

Some possible rules for strict mode, if we want to keep it:

* At the end of a job, `configuration`, `data` and ` references` are removed (are there any other managed props?)
* * Removing data is kind of a big deal, but at least is forces you to process response data in that job and prepare it for passing down the line.
* * It would also mean adaptors don't have to tidy up after themselves
* All properties beginning with `_` are removed (including on nested structures)
* Introduce a new property on state, maybe `result` or `final` or `persist`, which is the only thing that gets returned or forwarded.

**Comment by taylordowns2000** [comment]
@josephjclark , I don't know if this helps, but I _like_ the fuzziness of non-strict-mode because it accounts for two major use cases: (1) where `state.data` does exactly what we (I _think_ you too) want, and (2) where it doesn't.

## Thing 1
Consider this common pattern where each operation is from a different adaptor. (i'll use namespaced operations—which aren't a thing) to indicate the adaptor)
```js
OpenMRS.get('/patients/007');
DHIS2.post('/trackedEntityInstances', dataValue('patient')); 
```
Above, it makes perfect sense that the response from openMRS "get" is stored in data, and dhis2 "post" uses ONLY what's in `state.data` (this is why `dataValue()` is provided as a helper) and in `state.configuration`. All good. All `state.data`. Happy days.

## Thing 2
Then consider this _other common pattern_ in which the author wants to "pin something" to state so they can access it any time they want and know that they are in full control of what that pinned value (or function) is.
```js
OpenMRS.get('/patients/007', state => { ...state, patient: state.data });
DHIS2.post('/trackedEntityInstances', state => state.patient); 
HAPI-FHIR.post('/patients', state => state.patient);
```

In this usage, it's _almost_ as if we're looking at the DHIS2 post and the HAPI-FHIR post as being parallel. (In fact you could run them in parallel because they don't care about each other's data.) Setting aside parallelization, what's the final `state.data` in this flow? It's the response of the HAPI-FHIR server when we add a new patient record. Where's the DHIS2 response? It's in the `state.references` array.

This is all _OK_ (just fine?) by me, but I still sense that we _could_ be doing something better with the contract between operations and jobs. I think that's your "I kinda wish..." and I'll noodle on it for a while. Thanks for opening this issue ❤️

**Comment by josephjclark** [comment]
I think the sensible thing right now is to run in unstrict/legacy mode by default while we noodle. It's clearly not helping anyone.

I agree that everything there is fine :)

Thing 1 is great - my assumption is that you'd normally want to do some kind of transformation on the response. But maybe that's not the case?

If we were to enforce a stricter contract, you could consider this:

OpenMRS.get('/patients/007', persist('patient'));
DHIS2.post('/trackedEntityInstances', get('patient')); 

Where persist writes `patient` from `state.data` to, I don't know, `state.result`. And `get` reads from `state.result`.

But I'm just re-writing adaptor code for the sake of it now. I don't see a compelling reason to change yet.

Those same persist/get helpers would make Thing 2 a bit cleaner. There's a separate issue open for the parallelisation stuff so I won't touch that side of things. Otherwise I don't know that a stricter state contract would make it any easier.

**Comment by taylordowns2000** [comment]
I really like "persist" (or "pin"?) as a way of writing to a part of state that only you control.

> Thing 1 is great - my assumption is that you'd normally want to do some kind of transformation on the response. But maybe that's not the case?

For this 👆 , yeah... you're right. More common would be either (a)
```js
OpenMRS.get('/patients/007', state => {
  // ...manipulate and return openMRS patient
});
DHIS2.post('/trackedEntityInstances', dataValue('patient')); 
```
or b:
```js
OpenMRS.get('/patients/007');
Common.fn(state => { manipulateOpenMRSpatient });
DHIS2.post('/trackedEntityInstances', dataValue('patient')); 
```
or c:
```js
OpenMRS.get('/patients/007');
DHIS2.post('/trackedEntityInstances', {
  fullName: dataValue('patient.name'),
  age: Date.now() - dataValue('patient.dob')
}); 
```
There are different reasons for doing any of these. All are valid and have their place, IMO.

But yeah, I like "persist" or "pin". Let's keep noodling 🍝 🧠 🍜

**Comment by josephjclark** [comment]
Lightly related, but  an input to this "let's rethink how state is used" thing. These are not fully formed thoughts, but notes inspired by some stuff going on in language-sftp.

Some adaptors, like postgresql, write adaptor specific stuff to state, for internal use only. Postgres writes a `client` object to state, and adaptor functions use this magical state property to do their business (ie, CRUD functions in the adaptor pull the client off of state and use it to speak to the database).

As a code author, I return from an operation `{ data: myData }`, I will be inadvertently removing the client object from state and so breaking the job from that point on. Easy to recover from if you know what you're doing.

But the point is: as a job author you are not in control of your state object. Not only might the runtime have an opinion about how to process your state, but the adaptor may also have opinions on how state should look (in addition to the `data` property).

What I'm looking for in this strict mode stuff is a some really clean, clear separation of how the state object should be used. What keys are managed, what are added. If I return `{ data }`, will anything be restored before the next operation?

An approach like "whatever the operation returns will be written to `state.data`" sort of means that the author can only manipulate that piece of state, and the rest is managed. Which makes it harder to break an adaptor or accidentally print sensitive configuration.

**Comment by josephjclark** [comment]
A lot of this discussion is still relevant - but strict mode will be removed from the 1.0 runtime, so the issue is kinda moot.
